### PR TITLE
[EMB-342] Developer apps

### PR DESCRIPTION
**Pull request by aaxelb**
(based on #387 -- merge that first or review these together)

## Purpose
Add developer apps page

## Summary of Changes
- Add developer apps settings page
- Add `validated-model-form` component, use it for editing/creating developer apps and personal access tokens

Create:
developer app -- create

Edit:
developer app -- edit

Delete:
developer app -- delete

Reset client secret:
developer app -- reset

## Side Effects
Changed some internals of the tokens page, should be no visible effect.

## Feature Flags
`ember_user_settings_apps_page`

## QA Notes
- Create, edit, delete, reset developer apps at `/settings/applications`
- Validation FYI:
  - Name/URLs: required, max 200 characters
  - Description: optional, max 1000 characters
  - URLs need to be valid URLs, but no other checking is done
- Check that resetting a client secret actually results in a new secret
- Worth a quick regression of the tokens create/edit forms, just in case

## Ticket

https://openscience.atlassian.net/browse/EMB-342

# Reviewer Checklist

- [ ] meets requirements
- [ ] easy to understand
- [ ] DRY
- [ ] testable and includes test(s)
- [ ] changes described in `CHANGELOG.md`

### CODE REVIEW

File: mirage/views/developer-app.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+export function createDeveloperApp(this: HandlerContext, schema: Schema) {
+    const attrs = {
+        ...this.normalizedRequestAttrs(),
+        clientId: faker.internet.ip(),
+        clientSecret: faker.random.uuid(),
+    };
+    return schema.developerApps.create(attrs);
+}
+
+export function developerAppReset(this: HandlerContext, schema: Schema, request: Request) {
```
Comment (jamescdavis) [inline mirage/views/developer-app.ts]: `s/developerAppReset/resetClientSecret/`?

File: mirage/views/osf-resource.ts [hunk 2]
```diff
@@ -39,7 +40,10 @@ export function osfResource(
 
     if (actions.includes('index')) {
         server.get(opts.path, function(schema, request) {
-            const models = schema[modelName]
+            const collection = schema[modelName];
+            assert(`Unrecognized model name: ${modelName}`, Boolean(collection));
```
Comment (jamescdavis) [inline mirage/views/osf-resource.ts]: :heart:

File: tests/integration/components/validated-model-form/component-test.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        // Template block usage:
+        await render(hbs`
+            {{#validated-model-form}}
+                template block text
+            {{/validated-model-form}}
+        `);
+
+        assert.dom(this.element).hasText('template block text');
+    });
+});
```
Comment (jamescdavis) [inline tests/integration/components/validated-model-form/component-test.ts]: We should make this a real test as some point, but, for now...

File: app/settings/developer-apps/edit/template.hbs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        {{fa-icon 'arrow-left'}}
+        {{t 'settings.developer-apps.backToList'}}
+    {{/link-to}}
+</div>
+
+<h4>{{t 'settings.developer-apps.appDetail' appName=this.developerApp.name}}</h4>
+
+{{#if model.taskInstance.isError}}
+    {{model.taskInstance.error}}
+{{else}}
```
[the remark this answers is not in the document]
Comment (jamescdavis) [reply app/settings/developer-apps/edit/template.hbs]: Or maybe `validated-input`?

File: app/serializers/osf-serializer.ts [hunk 5]
```diff
@@ -1,8 +1,12 @@
 import { camelize, underscore } from '@ember/string';
-import DS from 'ember-data';
+import DS, { ModelRegistry } from 'ember-data';
 
 const { JSONAPISerializer } = DS;
 
+const API_TYPE_KEYS: Record<string, string> = {
+    applications: 'developer-app',
```
[the remark this answers is not in the document]
Comment (aaxelb) [reply app/serializers/osf-serializer.ts]: Yeah, I don't like it either, but `application` already means something else in some relevant namespaces (adapter, serializer), so we have to handle this somehow.

File: app/serializers/osf-serializer.ts [hunk 6]
```diff
@@ -111,12 +119,13 @@ export default class OsfSerializer extends JSONAPISerializer.extend({
...
 
-    serialize(snapshot: DS.Snapshot, options: object): any {
+    serialize(snapshot: DS.Snapshot, options?: { osf?: OsfSerializerOptions }): any {
         const serialized = this._super(snapshot, options);
         serialized.data.type = underscore(serialized.data.type);
 
-        // Only send dirty attributes and relationships in request
-        if (!snapshot.record.get('isNew')) {
+        const includeCleanData = options && options.osf && options.osf.includeCleanData;
+        if (!includeCleanData && !snapshot.record.get('isNew')) {
```
[the remark this answers is not in the document]
Comment (aaxelb) [reply app/serializers/osf-serializer.ts]: Yeah, I agree. I added a note to the ticket.

File: app/settings/developer-apps/edit/template.hbs (hunk 4 above)
[the remark this answers is not in the document]
Comment (aaxelb) [reply app/settings/developer-apps/edit/template.hbs]: The loading state is the form [with all input controls disabled](https://github.com/CenterForOpenScience/ember-osf-web/pull/392/files/fd10ebec938ee9916636996de009b9359940e6b6#diff-ea5557ac2d3460bb6757f41a833c8bacR35).

File: app/settings/developer-apps/x-app-card/component.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+@tagName('') // No div
+export default class DeveloperAppCard extends Component {
+    @service i18n!: I18n;
+    @service router!: RouterService;
+    @service toast!: Toast;
+
+    // Required arguments
+    developerApp!: DeveloperApp;
+    @requiredAction deleted!: () => unknown;
```
[the remark this answers is not in the document]
Comment (aaxelb) [reply app/settings/developer-apps/x-app-card/component.ts]: Went with `onDelete`

File: app/serializers/osf-serializer.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (brianjgeiger) [reply app/serializers/osf-serializer.ts]: Would it be worth making an `enum` where the keys are the types as they should be, and the values are the what we need them to be for reasons?

File: app/locales/en/translations.ts [hunk 8]
```diff
@@ -1137,8 +1140,38 @@ export default {
         notifications: {
             title: 'Notifications',
         },
-        apps: {
+        'developer-apps': {
```
[the remark this answers is not in the document]
Comment (aaxelb) [reply app/locales/en/translations.ts]: Hmm, I'm not sure. Longer keys aren't ideal. I understand the desire to impose order, but I don't know if splitting up routes and components will make much difference.

**Pull request merged**